lib/providers/base-provider.js

```diff
@@ -0,0 +1,80 @@
+var util = require('util'),
+    EventEmitter = require('events').EventEmitter;
+
+// Represents the base class for all providers
+function BaseProvider() {
+  // derived children should start connection and authentication logic
+  // in the constructor, and emit 'open' or 'close' as a result of success 
+  // or failure. backoff logic should also be implemented.
+};
+
+util.inherits(BaseProvider, EventEmitter);
+
+// Should be overriden in providers
+// 
+// expects jsonData:string - a string of json to write to the provider
+BaseProvider.prototype.write = function(jsonData) {
+  throw new Error("Providers should override BaseProvider.prototype.write");
+};
+
+// This is the method that ConMan calls to send data to a provider
+// it will check if messages are of the right type, and if so, write
+// them to the provider as JSON.
+// 
+// This should not be overriden by providers
+// 
+// expects (ProviderMessage)
+// or (ProviderMessage[]) or (ProviderMessage 1, ProviderMessage 2, ProviderMessage N)
+// as argument(s)
+BaseProvider.prototype.send = function() {
+  var args = Array.prototype.slice.call(arguments);
+
+  // if we're passed an array of ProviderMessage's, extract it
+  if (args.length == 1 && args[0] instanceof Array) {

```

Comment:
You check if `args[0]` is defined here, but on line 38 you don't. Not sure if there is an edge case where `args[0]` is null or undefined, but I'm assuming there is since you're checking here.